src/main/java/com/example/gymApp/service/TrainerService.java

```diff
@@ -1,131 +1,167 @@
+
+  public Trainer createTrainer(String firstName, String lastName, String username, String password,
+      String specialization) {
+    if (userRepository.findByUsername(username).isPresent()) {
+      throw new IllegalArgumentException("Username of trainer already exists");
     }
+
+    TrainingType trainingType = checkSpecializationCorrectness(specialization);
+
+    User user = new User();
+    user.setFirstName(firstName);
+    user.setLastName(lastName);
+    user.setUsername(username);
+    user.setPassword(password);
+    user.setActive(true);
+
+    Trainer trainer = new Trainer(user, trainingType);
+
+    trainerRepository.save(trainer);
+
+    log.info("Trainer created successfully: {}", trainer.getUser().getUsername());
     return trainer;
   }
 
-  public List<Trainer> getAllTrainers() {
-    try {
-      List<Trainer> trainers = trainerDAO.getAllTrainers();
-      log.info("Successfully retrieved list of trainers.");
-      return trainers;
-    } catch (Exception e) {
-      log.error("Error fetching all trainers: {}", e.getMessage());
-      throw new RuntimeException("Unable to retrieve trainers.");
-    }
+
+  public Trainer getTrainerByUsername(String username) {
+    return trainerRepository.findByUserUsername(username)
+        .orElseThrow(
+            () -> new NoSuchElementException("No trainer found with the username: " + username));
   }
 
-  private void validateTrainer(Trainer trainer) {
-    if (trainer.getId() == null || trainer.getId() <= 0) {
-      throw new IllegalArgumentException("Invalid ID: " + trainer.getId());
+  public User getTrainerByPassword(String password) {
+
+    Optional<User> userOpt = userRepository.findByPassword(password);
+
+    if (userOpt.isEmpty()) {
+      throw new IllegalArgumentException("No user found with the provided password");
     }
-    if (trainer.getFirstName() == null || trainer.getFirstName().trim().isEmpty()) {
-      throw new IllegalArgumentException("First name cannot be empty.");
+
+    User user = userOpt.get();
+
+    return user;
+
+
+  }
+
+  @Transactional
+  public void deleteTraineeByUsername(String username) {
+
+    Optional<User> userOpt = userRepository.findByUsername(username);
+
+    if (userOpt.isEmpty()) {
+      throw new NoSuchElementException("No user found with the provided username");
     }
-    if (trainer.getLastName() == null || trainer.getLastName().trim().isEmpty()) {
-      throw new IllegalArgumentException("Last name cannot be empty.");
+
+    User user = userOpt.get();
+
+    Optional<Trainee> traineeOpt = traineeRepository.findByUser(user);
+
+    if (traineeOpt.isEmpty()) {
+      throw new NoSuchElementException("No trainee found for the provided user");
     }
-    if (trainer.getSpecialization() == null || trainer.getSpecialization().trim().isEmpty()) {
-      throw new IllegalArgumentException("Specialization cannot be empty.");
+    Trainee trainee = traineeOpt.get();
+
+    //deleting connected entities manually!!!
```

Comment:
In the task mentioned: and bring the cascade deletion of relevant trainings. Point 7 in the notes.